VizLegend: Represent line style in series legend and tooltip

domasx2 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

This PR makes panel legend & viz tooltip  color indicator also represent line style of a series: solid, dotted or dashed. Only works if gradient is not set, because we can't represent it in dashed/ dotted form with current appraoch.

![dot-dash-cmp](https://github.com/grafana/grafana/assets/847684/da62ef7e-660f-4986-b8c4-09569f7ac021)
![dot-dash-cmp](https://github.com/grafana/grafana/assets/847684/781d4cf0-f666-496d-ba8d-cf0db65e20e3)

**Why do we need this feature?**

This is especially useful with time comparison feature where UX calls for comparison series to have same color but different line style. Eg, app o11y uses dashed lines for previous period series. Without legend representing line style it is harder to distinguish legend items.

**Who is this feature for?**

Dashboard users, plugins.


**Special notes for your reviewer:**

I went ahead and made viz tooltip reuse `SeriesIcon` component for color indicator

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

domasx2 committed:
wip

domasx2 committed:
aand finished

domasx2 committed:
remove console log

leeoniya commented:
would it be possible to avoid extra DOM here? especially for tooltip that has to re-render up to 60 times per sec, this will add a lot of overhead for something that we might be able to do with a single SVG element.

domasx2 commented:
> would it be possible to avoid extra DOM here? especially for tooltip that has to re-render up to 60 times per sec, this will add a lot of overhead for something that we might be able to do with a single React.memo'd SVG element or clever CSS.

I'll try. What's the case where dom is recreated 60 times per second though? :-O would React.memo'ing the divs work too?

leeoniya commented:
> I'll try. What's the case where dom is recreated 60 times per second though?

sorry, i usually mean react's v-tree, which very unfree to re-create on ever mousemove. imagine you have 30 series in a tooltip. this adds up very fast.

domasx2 committed:
use css for styling

domasx2 commented:
I removed extra divs and refactored with border styling, which gives almost the same result in Chrome 
![image](https://github.com/grafana/grafana/assets/847684/e2039682-e165-4908-a86f-d3e78cc0e3e2)

Will test with more browsers tomorrow

leeoniya commented on packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx:
```diff
@@ -29,28 +31,47 @@ export const SeriesIcon = React.memo(
+    width: '14px',
+  }),
+  dot: css({
+    borderTop: 'dotted 3.99px',
```
3.99, huh? :sweat_smile:

domasx2 replied:
it feels a lot more affordable than 4 :D 

using `4` makes only 2 dots show up with a large gap *on my machine*, hence the odd "hack"

leeoniya commented:
> I removed extra divs and refactored with border styling, which gives almost the same result in Chrome

thanks :)

leeoniya commented on packages/grafana-ui/src/components/VizLegend/SeriesIcon.tsx:
```diff
@@ -29,28 +31,47 @@ export const SeriesIcon = React.memo(
+    const lineStyleStyle =
+      lineStyle?.fill === 'dot' ? styles.dot : lineStyle?.fill === 'dash' ? styles.dash : styles.solid;
+
+    const colorStyle: CSSProperties =
+      lineStyle?.fill === 'dot' || lineStyle?.fill === 'dash'
+        ? {
+            borderTopColor: cssColor,
+          }
+        : {
+            backgroundColor: cssColor,
+          };
```
we can probably use a fat border for all cases (even solid)?

domasx2 replied:
done!

domasx2 committed:
user boerder for solid as well

domasx2 committed:
misc fixes

domasx2 commented:
Problem with border based solution is that it's not consistent across render engines. In firefox: 
![image](https://github.com/grafana/grafana/assets/847684/637a620e-e00d-4280-9d11-7c665dd234ab)

domasx2 committed:
attempt css based approach

domasx2 commented:
So border did not work out, too inconsistent across borwsers.
Refactored to third attempt which seems to work well:  repeated background images using `radial-gradient` and `linear-gradient`. This gives consistent result across all browsers.
Caveat is that it doesn't render gradient too, so in case gradient is set it will default to previous looks

leeoniya committed:
Merge branch 'main' into domas-line-style-legend

leeoniya reviewed (approved):
awesome! :heart: :shipit:
